### PR TITLE
[bitnami/mysql] New major version

### PR DESCRIPTION
**Pull request by juan131**
Signed-off-by: juan131 <juanariza@vmware.com>

**Description of the change**

This PR refactors the whole chart to follow Helm chart best practices. This includes several breaking changes, so please review it carefully, putting special emphasis to the "Upgrading" section describing all of them.

The most important change are:

- The terms *master* and *slave* have been replaced by the terms *primary* and *secondary*.
- Labels are adapted to follow the [Helm charts standard labels](https://helm.sh/docs/chart_best_practices/labels/#standard-labels).

**Benefits**

- Standardization.
- Reduces technical debt.

**Possible drawbacks**

It breaks backwards compatibility.

**Applicable issues**

None

**Additional information**

Upgrade tests from version `7.x.x` using workaround suggested in the Upgrading Notes:

- Install old chart version:

```console
$ helm install mysql bitnami/mysql \
    --set root.password=root_password \
    --set replication.enabled=true \
    --set replication.password=rep_password \
    --set db.user=my_username \
    --set db.name=my_database \
    --set db.password=my_password \
    --version 7.1.0
```

- Add some data once everything is up and running:

```console
$ kubectl run mysql-client --rm --tty -i --restart='Never' --image  docker.io/bitnami/mysql:8.0.22-debian-10-r23 --namespace default --command -- bash
I have no name!@mysql-client:/$ mysql -h mysql.default.svc.cluster.local -uroot -p my_database
Enter password:
...
mysql> CREATE TABLE foo (a int);
Query OK, 0 rows affected (0.04 sec)
mysql> INSERT INTO foo (a) VALUES (10);
Query OK, 1 row affected (0.01 sec)
```

- Remove chart and slave PVC (we only need the master PVC since slave only replicates data):

```console
$ helm uninstall mysql && kubectl delete pvc data-mysql-slave-0
$ kubectl get pvc
NAME                  STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-mysql-master-0   Bound    pvc-d39cdea9-5371-4445-b11b-eed187d06daa   8Gi        RWO            standard       2m45s
```

- Install new chart version reusing existing PVC for primary node:

```console
$ helm install mysql bitnami/mysql \
  --set architecture=replication \
  --set auth.rootPassword=root_password \
  --set auth.replicationPassword=rep_password \
  --set auth.username=my_username \
  --set auth.password=my_password \
  --set auth.database=my_database \
  --set primary.persistence.existingClaim=data-mysql-master-0 \
  --version 8.0.0
```

- Ensure the data is reused/upgraded:

```console
$ kubectl logs mysql-primary-0
...
mysql 15:40:56.97 INFO  ==> Using persisted data
mysql 15:40:57.00 INFO  ==> Running mysql_upgrade
...
```

- Check data is okay:

```console
$ kubectl run mysql-client --rm --tty -i --restart='Never' --image  docker.io/bitnami/mysql:8.0.22-debian-10-r23 --namespace default --command -- bash
I have no name!@mysql-client:/$ mysql -h mysql-primary.default.svc.cluster.local -uroot -p my_database
Enter password:
...
mysql> SELECT * FROM foo;
+------+
| a    |
+------+
|   10 |
+------+
1 row in set (0.02 sec)
mysql> SHOW GRANTS FOR 'my_username';
+--------------------------------------------------------------+
| Grants for my_username@%                                     |
+--------------------------------------------------------------+
| GRANT USAGE ON *.* TO `my_username`@`%`                      |
| GRANT ALL PRIVILEGES ON `my_database`.* TO `my_username`@`%` |
+--------------------------------------------------------------+
2 rows in set (0.00 sec)
```

**Checklist**

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [x] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

### CODE REVIEW

File: bitnami/mysql/templates/_helpers.tpl [hunk 1]
```diff
@@ -100,232 +46,104 @@ Get the initialization scripts ConfigMap name.
...
-{{- $messages := append $messages (include "mysql.validateValues.loadBalancerIPareNotEquals" .) -}}
-{{- $messages := without $messages "" -}}
-{{- $message := join "\n" $messages -}}
-
-{{- if $message -}}
-{{-   printf "\nVALUES VALIDATION:\n%s" $message | fail -}}
+{{- define "mysql.createSecret" -}}
+{{- if not .Values.auth.existingSecret }}
+    {{- true -}}
+{{- else -}}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/_helpers.tpl:127]: This else is not necessary, by default it returns nothing.

File: bitnami/mysql/templates/_helpers.tpl [hunk 2]
```diff
@@ -100,232 +46,104 @@ Get the initialization scripts ConfigMap name.
...
+{{- $messages := without $messages "" -}}
+{{- $message := join "\n" $messages -}}
 
-{{/*
-Returns chart secret name. If existingSecret is not set it will default to mysql.fullname
-*/}}
-{{- define "mysql.secretName" -}}
-{{ default (include "mysql.fullname" .) .Values.existingSecret }}
+{{- if $message -}}
+{{-   printf "\nVALUES VALIDATION:\n%s" $message | fail -}}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/_helpers.tpl:147]:
```suggestion
{{- printf "\nVALUES VALIDATION:\n%s" $message | fail -}}
```

File: bitnami/mysql/templates/primary/configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    app.kubernetes.io/component: primary
+    {{- if .Values.commonLabels }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonLabels "context" $ ) | nindent 4 }}
+    {{- end }}
+  {{- if .Values.commonAnnotations }}
+  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+  {{- end }}
+data:
+  my.cnf: |-
+{{ .Values.primary.configuration | indent 4 }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/primary/configmap.yaml:17]:
```suggestion
    {{ .Values.primary.configuration | nindent 4 }}
```

Just for consistency

File: bitnami/mysql/templates/primary/statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+          imagePullPolicy: {{ .Values.metrics.image.pullPolicy | quote }}
+          env:
+            {{- if .Values.auth.usePasswordFiles }}
+            - name: MYSQL_ROOT_PASSWORD_FILE
+              value: "/opt/bitnami/mysqld-exporter/secrets/mysql-root-password"
+            {{- else }}
+            - name: MYSQL_ROOT_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: {{ template "mysql.secretName" . }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/primary/statefulset.yaml:235]:
```suggestion
                  name: {{ include "mysql.secretName" . }}
```

File: bitnami/mysql/templates/primary/statefulset.yaml [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+      volumes:
+        {{- if or .Values.primary.configuration .Values.primary.existingConfigmap }}
+        - name: config
+          configMap:
+            name: {{ include "mysql.primary.configmapName" . }}
+        {{- end }}
+        {{- if or .Values.initdbScriptsConfigMap .Values.initdbScripts }}
+        - name: custom-init-scripts
+          configMap:
+            name: {{ template "mysql.initdbScriptsCM" . }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/primary/statefulset.yaml:283]:
```suggestion
            name: {{ include "mysql.initdbScriptsCM" . }}
```

File: bitnami/mysql/templates/primary/statefulset.yaml [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        {{- end }}
+        {{- if .Values.auth.usePasswordFiles }}
+        - name: mysql-credentials
+          secret:
+            secretName: {{ template "mysql.secretName" . }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/primary/statefulset.yaml:288]:
```suggestion
            secretName: {{ include "mysql.secretName" . }}
```

File: bitnami/mysql/templates/primary/statefulset.yaml [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+                path: mysql-replication-password
+              {{- end }}
+        {{- end }}
+        {{- if .Values.primary.extraVolumes }}
+        {{- include "common.tplvalues.render" (dict "value" .Values.primary.extraVolumes "context" $) | nindent 8 }}
+        {{- end }}
+  {{- if and .Values.primary.persistence.enabled .Values.primary.persistence.existingClaim }}
+        - name: data
+          persistentVolumeClaim:
+            claimName: {{ tpl .Values.primary.persistence.existingClaim . }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/primary/statefulset.yaml:305]: Are we using `tpl` for any reason?

File: bitnami/mysql/templates/secondary/configmap.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    app.kubernetes.io/component: secondary
+    {{- if .Values.commonLabels }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonLabels "context" $ ) | nindent 4 }}
+    {{- end }}
+  {{- if .Values.commonAnnotations }}
+  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+  {{- end }}
+data:
+  my.cnf: |-
+{{ .Values.secondary.configuration | indent 4 }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/secondary/configmap.yaml:17]:
```suggestion
    {{ .Values.secondary.configuration | nindent 4 }}
```

File: bitnami/mysql/templates/secondary/statefulset.yaml [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+        app.kubernetes.io/component: secondary
+        {{- if .Values.commonLabels }}
+        {{- include "common.tplvalues.render" ( dict "value" .Values.commonLabels "context" $ ) | nindent 8 }}
+        {{- end }}
+    spec:
+      {{- include "mysql.imagePullSecrets" . | nindent 6 }}
+      {{- if .Values.schedulerName }}
+      schedulerName: {{ .Values.schedulerName | quote }}
+      {{- end }}
+      serviceAccountName: {{ template "mysql.serviceAccountName" . }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/secondary/statefulset.yaml:48]:
```suggestion
      serviceAccountName: {{ include "mysql.serviceAccountName" . }}
```

File: bitnami/mysql/templates/secondary/statefulset.yaml [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+            {{- if .Values.auth.usePasswordFiles }}
+            - name: MYSQL_REPLICATION_PASSWORD_FILE
+              value: "/opt/bitnami/mysql/secrets/mysql-replication-password"
+            {{- else }}
+            - name: MYSQL_REPLICATION_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: {{ template "mysql.secretName" . }}
+                  key: mysql-replication-password
+            {{- end }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/secondary/statefulset.yaml:138]: It is the same condition maybe we can merge the content in a single `if-else` block

File: bitnami/mysql/templates/secrets.yaml [hunk 11]
```diff
@@ -1,34 +1,39 @@
...
+    {{- end }}
+  {{- if .Values.commonAnnotations }}
+  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+  {{- end }}
 type: Opaque
 data:
-  {{- if .Values.root.password }}
-  mysql-root-password: {{ .Values.root.password | b64enc | quote }}
-  {{- else if (not .Values.root.forcePassword) }}
+  {{- if not (empty .Values.auth.rootPassword) }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/secrets.yaml:16]: Just for curiousity, and empty string would not be evaluated to false?

File: bitnami/mysql/templates/serviceaccount.yaml [hunk 12]
```diff
@@ -2,15 +2,17 @@
...
-  labels:
-    app: {{ template "mysql.name" . }}
-    chart: {{ template "mysql.chart" .  }}
-    release: "{{ .Release.Name }}"
-    heritage: "{{ .Release.Service }}"
-  {{- if .Values.serviceAccount.annotations }}
-  annotations: {{ toYaml .Values.serviceAccount.annotations | nindent 4 }}
-  {{- end }}
-secrets:
-  - name: {{ template "mysql.secretName" . }}
```
Comment (miguelaeh) [inline bitnami/mysql/templates/serviceaccount.yaml:15]: It seems that this section is dissapearing

**Review by miguelaeh — COMMENTED** [review]
Good job!
I added some comments but in general LGTM!

File: bitnami/mysql/templates/secrets.yaml (hunk 11 above)
Comment (juan131) [reply bitnami/mysql/templates/secrets.yaml:16]: Yes, but this explicitly states it.

File: bitnami/mysql/templates/primary/statefulset.yaml (hunk 7 above)
Comment (juan131) [reply bitnami/mysql/templates/primary/statefulset.yaml:305]: In case the user want to use a template in the name

**Review by miguelaeh — APPROVED** [review]
LGTM!

**Pull request merged**